File: ScriptTester/src/main/scala/ScriptTester.scala
import java.io.File
import collection.mutable.Buffer
import scala.collection.JavaConversions._

object ScriptTester {
  var srcnum = 0
  var done = false
  def main(args : Array[String]) {
    println("Hello! This is a world of scala!!!")
    var mangaManager = MangaManager.instance()
    initFolders()
    ScriptManager.init()
    sourceLoop()
    //printSources() 
    //setSourceNumber(readInt)
    //println("You chose source number "+srcnum)
    //changeSource()
    /*var list = getChapterList()*/
    //while(!done) {
    //printMangaList(list)
    //println("Next page (n), Previous page (p), Change Source (c), Quit (q), or Manga Number:")
    //val ln = readLine()
    //if(ln(0) == 'q') {
      //println("Exiting!!")
      //done = true
    //} else if(ln(0) == 'n') {
      //list = getSourceNextPage()
    //} else if(ln(0) == 'p') {
      //list = getSourcePreviousPage()
    //} else if(ln(0) == 'c') {
      //changeSource()
      //list = getChapterList()
    //} else {
      //println("you chose the manga "+list(ln.toInt).getTitle)
      //ScriptManager.setCurrentSource(srcnum)
      //MangaManager.readingOffline(false)
      //MangaManager.setCurrentManga(list(ln.toInt))
      //done = true
    //}
  /*}*/
  }
  
  def sourceLoop() = {
    var sloop = true
    while(sloop) {
      changeSource()
      mangaLoop()
    }
  }

  def mangaLoop() = {
    var mloop = true
    var list = getMangaList()
    while(mloop) {
      printMangaList(list)
      println("Next page (n), Previous page (p), Back (b), Quit (q), or Manga Number:")
      val ln = readLine()
      if(ln(0) == 'q') {
        println("Exiting!!")
        System.exit(0)
      } else if(ln(0) == 'n') {
        list = getSourceNextPage()
      } else if(ln(0) == 'p') {
        list = getSourcePreviousPage()
      } else if(ln(0) == 'b') {
        mloop = false
        //changeSource()
        //list = getChapterList()
      } else {
        var num = ln.toInt
        println("you chose the manga "+list(num).getTitle)
        ScriptManager.setCurrentSource(srcnum)
        MangaManager.readingOffline(false)
        MangaManager.setCurrentManga(list(num))
        chapterLoop()
      }
    }
  }

  def chapterLoop() = {
    var cloop = true
    val manga = MangaManager.getCurrentManga()
    var list = MangaManager.getMangaChapterList()
    println("Description: "+manga.getDescription())
    while(cloop){
      printChapterList(list)
      println("Back (b), Quit (q), or Chapter Number:")
      val ln = readLine()
      if(ln(0) == 'q') {
        println("Exiting!!")
        System.exit(0)
      } else if(ln(0) == 'b') {
        cloop = false
      } else {
        var num = ln.toInt
        println("you chose the chapter "+list(num).getTitle)
        MangaManager.setCurrentChapter(list(num).asInstanceOf[Chapter])
        MangaManager.setCurrentPageNum(0)
        println("Debug String 1")
        imageLoop()
      }

    }
  }
  
  def imageLoop() = {
    var iloop = true
    println("Debug String 2")
    var list : Buffer[String] = Buffer[String]()
    var current : String = null
    var total = MangaManager.getNumPages()
    println("Debug String 3")
    current = updateImage()
    list.append(current)
    println("Debug String 4")
    while(iloop) {
      println("Images downloaded: ")
      printImageList(list)
      println("Back (b), Quit (q), Next image (n), Previous Image (p):")
      val ln = readLine()
      if(ln(0) == 'q') {
        System.exit(0)
      } else if(ln(0) == 'b') {
        iloop = false
      } else if(ln(0) == 'n') {
        //get next image!
        var i = MangaManager.getCurrentPageNum()
        if(i < total-1) {
          MangaManager.setCurrentPageNum(i+1)
        } else {
          if(MangaManager.nextChapter()) {
            MangaManager.setCurrentPageNum(0)
          }
        }
        current = updateImage()
        list.append(current)
      } else if(ln(0) == 'p') {
        //get previoust image!
        var i = MangaManager.getCurrentPageNum()
        if(i > 0) {
          MangaManager.setCurrentPageNum(i-1)
        } else {
          if(MangaManager.previousChapter()) {
            MangaManager.setCurrentPageNum(MangaManager.getNumPages - 1)
          }
        }
        current = updateImage()
        list.append(current)
      } else {
        println("Error, unrecognized command!")
      }
    }
  }

  def initFolders() = {
    println("Making folders")  
    var mainfolder = new File("./Mangagaga")
    try {
      if(!mainfolder.exists) {
        mainfolder.mkdir
        for (foldername <- List("Downloaded", "Scripts", "Cache")) {
          val folder = new File(mainfolder, foldername)
          if(!folder.exists) {
            folder.mkdir
          }
        }
      }
    } catch {
      case e:Exception => System.out.println("There was an error making folders!")
    }
  }
  
  def setSourceNumber(num: Int) = {
    if(num < 0) {
      println("Exiting")
    } else {
      srcnum = num
    }
  }
  
  def printSources() = {
    println("\nSelect a source")
    for(i <- 0 until ScriptManager.numSources) {
      println(i+": "+ScriptManager.getScript(i).getName)
    }
  }
  
  def changeSource() = {
    printSources() 
    setSourceNumber(readInt)
    println("You chose source number "+srcnum)
  }
  
  def printMangaList(list : Buffer[Manga]) = {
    var count = 0
    for(i <- list) {
      println(count + ": "+i.getTitle)
      count += 1
    }
  }
  
  def printChapterList(list : Buffer[Chapter]) = {
    var count = 0
    for(i <- list) {
      println(count + ": "+i.getTitle)
      count += 1
    }
  }

  def printImageList(list : Buffer[String]) = {
    var count = 0
    for(i <- list) {
      println(count + ": "+i)
      count += 1
    }
  }
  
  def updateImage() : String = {
    MangaManager.getCurrentPage()
  }

  def getMangaList() : Buffer[Manga] = {
    ScriptManager.getScript(srcnum).getMangaListPage1()
  }

  def getSourceNextPage() : Buffer[Manga] = {
    ScriptManager.getScript(srcnum).getMangaListNextPage()
  }

  def getSourcePreviousPage() : Buffer[Manga] = {
    ScriptManager.getScript(srcnum).getMangaListPreviousPage()
  }
}

File: tester/src/main/kotlin/ScriptTester.kt
package io.githup.limvot.mangagaga

import java.awt.FlowLayout
import java.io.File
import javax.imageio.ImageIO
import javax.swing.ImageIcon
import javax.swing.JFrame
import javax.swing.JLabel

object ScriptTester {
    @JvmStatic
    fun main(vararg args : String) {
        SettingsManager.mangagagaPath = "Mangagaga"

        // make folders
        var mainfolder = File("./Mangagaga")
        try {
            if (!mainfolder.exists()) mainfolder.mkdir()
            for (foldername in listOf("Downloaded", "Scripts", "Cache")) {
                val folder = File(mainfolder, foldername)
                if (!folder.exists()) folder.mkdir()
            }
        } catch (e: Exception) {
            println("There was an error making folders!")
        }
        SettingsManager.loadSettings()
        // Overwrite all scripts
        Utilities.gitToScripts()

        Boss.init()
        var state = "source"
        while (true) { when (state) {
            "source"  -> state = sourceLoop()
            "manga"   -> state = mangaLoop()
            "chapter" -> state = chapterLoop()
            "image"   -> state = imageLoop()
        } }
    }

    fun sourceLoop() : String {
        val script_options = Boss.getScriptList()
        printList(script_options)
        println("\nSelect a source")
        Boss.currentSource = script_options[readLine()!!.toInt()]

        val types = Boss.getFilterTypes()
        printList(types)
        println("\nSelect a type")
        Boss.currentFilter = types[readLine()!!.toInt()]
        return "manga"
    }

    fun mangaLoop() : String {
        val ml_list = Boss.getMangaList()
        printList(ml_list)
        val ln = printPrompt("Back (b), Quit (q), or Manga Number:")
        if (ln[0] == 'b') return "source"
        Boss.currentManga = ml_list[ln.toInt()]
        return "chapter"
    }

    fun chapterLoop() : String {
        println("Description: ${Boss.getMangaDescription()}")
        val cl_list = Boss.getChapterList()
        printList(cl_list)
        val ln = printPrompt("Back (b), Quit (q), or Chapter Number:")
        if (ln[0] == 'b') return "manga"
        Boss.currentChapter = cl_list[ln.toInt()+1]
        Boss.currentPage = 0
        return "image"
    }

    fun printList(list : List<String>) = println(list.withIndex().map { (i,x) -> "$i: $x" }.joinToString("\n"))

    fun imageLoop(): String {
        //request number of pages
        println("Num Pages ${Boss.getNumPages()}")

        //request first page
        Boss.currentPage = 0
        var current = Boss.getPagePath()

        val il_frame = JFrame()
        il_frame.layout = FlowLayout()
        il_frame.setSize(800,1200)
        val il_lbl = JLabel()
        il_frame.add(il_lbl)
        il_frame.setVisible(true)
        il_frame.defaultCloseOperation = JFrame.EXIT_ON_CLOSE

        while (true) {
            il_lbl.icon = ImageIcon(ImageIO.read(File(current)))
            val ln = printPrompt("Back (b), Quit (q), Next image (n), Previous Image (p):")
            when (ln[0]) {
                'b'    -> return "chapter"
                'n','p'-> {
                    Boss.move((ln[0] == 'n'))
                    current = Boss.getPagePath();
                }
                else   -> println("Error, unrecognized command!")
            }
        }
    }

    fun printPrompt(msg : String) : String {
        println(msg)
        val ln = readLine()!!
        if (ln[0] == 'q') {
            println("Exiting!!")
            System.exit(0)
        }
        return ln
    }
}
